group/assignment6/RearrangeCars.cpp

```diff
@@ -0,0 +1,141 @@
+	currentStates[to] = currentStates[from];
+	currentStates[from] = -1;
+	carPosition[currentStates[to]] = to;
+}
+
+void RearrangeCars::generateSequenceOfMoves(const vector<int> &endStates, vector<Move> &sequenceOfMoves) {
+	vector<int> currentStates = this->startStates; // startStates will remain unchanged
+	vector<int> carPosition = findCarPositions(currentStates);
+
+	int emptySlotId = findEmptySlotId();
+	int currCarId, currCarSlotId;
+
+	while (true) {
+		currCarId = endStates[emptySlotId];
+		if (currCarId == -1)	// We reached slot which should be empty in the end
+			currCarId = findMisplacedCar(carPosition, endStates);
+		if (currCarId == -1) break;
+		currCarSlotId = carPosition[currCarId];
+		moveCar(currentStates, carPosition, currCarSlotId, emptySlotId);
+		sequenceOfMoves.push_back(Move(currCarId, make_pair(currCarSlotId, emptySlotId)));
+		emptySlotId = currCarSlotId;
+	}
+}
+
+bool RearrangeCars::validMove(int carId, int emptySlotId, const vector<set<int>>& constrains) {
+	if (constrains[emptySlotId].size() == 0)	// There are no constraints for this parking space, any car can be moved here
+		return true;
+
+	if ((constrains[emptySlotId]).find(carId) != (constrains[emptySlotId].end()))
+		return true;
+
+	return false;
+}
+
+int RearrangeCars::findMisplacedCarWithConstraints(vector<int> &carPosition, vector<int> &currentStates, const vector<int> &endStates, const vector<set<int>> &constraints, vector<Move> &sequenceOfMoves, int &emptySlotId, int &numOfCarsOnDisiredPosition) {
+	int lastMovedCarId = -1;
+	int currCarSlotId;
+
+	while (true) {
+		currCarSlotId = endStates[emptySlotId];
+		if (lastMovedCarId != currCarSlotId && currCarSlotId != -1) {
+			return currCarSlotId;
+		}
+
+		// Find a displaced car that can be moved here
+		bool foundDisplacedCar = false;
+		for (int carId = 0; carId < carPosition.size(); carId++) {
+			if (lastMovedCarId != carId && endStates[carPosition[carId]] != carId && validMove(carId, emptySlotId, constraints)) {
+				currCarSlotId = carPosition[carId];
+				lastMovedCarId = carId;
+
+				foundDisplacedCar = true;
+				break;
+			}
+		}
+
+		// If displaced car can't be moved to free slot, move a car that is on its position and can be moved here
+		if (!foundDisplacedCar) {
+			for (int carId = 0; carId < carPosition.size(); carId++) {
+				if (validMove(carId, emptySlotId, constraints)) {
+					currCarSlotId = carPosition[carId];
+					lastMovedCarId = carId;
+
+					numOfCarsOnDisiredPosition--;
+					break;
+				}
+			}
+		}
+
+		// Make a move
+		moveCar(currentStates, carPosition, currCarSlotId, emptySlotId);
+		sequenceOfMoves.push_back(Move(lastMovedCarId, make_pair(currCarSlotId, emptySlotId)));
+		emptySlotId = currCarSlotId;
+	}
+
+}
+
+int RearrangeCars::numberOfCarsOnDesiredPosition(const vector<int>& carPosition, const vector<int>& endStates) {
+	int num = 0;
+	for (int i = 0; i < carPosition.size(); i++)
+		if (endStates[carPosition[i]] == i)
+			num++;
+
+	return num;
+}

```

Comment:
Just for readability, I would replace "num" with "count" or "numOfCarsInDesiredPosition"